c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -2837,14 +2837,38 @@ void local_raw_delete(void* ptr);
 
 class NativeCachingAllocator : public CUDAAllocator {
  private:
-  std::mutex mutex;
+  static constexpr size_t kNumMutexShard = 1031;
+
+  // TODO: change to hardware_destructive_interference_size when C++17 is available.
+  struct alignas(64) AlignedMutex {
+    std::mutex m;
+  };
+
+  std::array<AlignedMutex, kNumMutexShard> mutex;
 
   // allocated blocks by device pointer
-  ska::flat_hash_map<void*, Block*> allocated_blocks;
+  std::array<ska::flat_hash_map<void*, Block*>, kNumMutexShard>
+      allocated_blocks;
+
+  static constexpr uint64_t twang_mix64(uint64_t key) noexcept {
```

Comment:
Is there a specific reason use this? Any builtin hash we could use here?

Reply:
the hash of an integer return the integer itself which is not helpful to balance the load

Reply:
I wasn't necessarily talking about std::hash but in dependencies we already have or even something from c10/util/hash.h ?

Thinking back at this, why isn't just doing `ptr % kNumMutexShard` would be unbalanced or maybe  `ptr >> 2 % kNumMutexShard` ? I guess our allocation lowest significant bits are pretty random.

Reply:
seems c10::hash fallback to std::hash? https://codebrowser.bddppq.com/pytorch/pytorch/c10/util/hash.h.html#258

Reply:
> Thinking back at this, why isn't just doing `ptr % kNumMutexShard` would be unbalanced or maybe `ptr >> 2 % kNumMutexShard` ? I guess our allocation lowest significant bits are pretty random.

I think we can config alloc size to round up to a certain size. In that case, the lowest significant bits could be quite similar.

Reply:
Ok,
Would the sha1 implementation from c10/util/hash.h be fast enough for your use case here?
If not, adding a twang mix as a new hash sounds ok but I think it would be best to add it in next to the sha1 so that we can reuse it next time we need it.

Reply:
sounds good! sha1 looks slow for its std::string interface. let me move the function to hash.h.